bam/merger.go

```diff
@@ -0,0 +1,187 @@
+// Copyright ©2017 The bíogo Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package bam
+
+import (
+	"container/heap"
+	"errors"
+	"io"
+
+	"github.com/biogo/hts/sam"
+)
+
+// Merger implements merging BAM data with a defined sort
+// order. It can be used for sorting, concatenating and
+// deduplicating BAM data.
+type Merger struct {
+	h *sam.Header
+	// refLinks is the set of mappings from a RefID in
+	// a src Header to a Reference in the dst Header.
+	refLinks [][]*sam.Reference
+
+	less    func(a, b *sam.Record) bool
+	readers []*reader
+}
+
+type reader struct {
+	id int
+	r  *Reader
+
+	head *sam.Record
+	err  error
+}
+
+// NewMerger returns a Merger that reads from the source
+// Readers.
+//
+// The sort order of the stream merge is defined by the sort
+// order field of the src Reader headers and the provided less
+// function. The header sort order fields must agree.
+//
+// Sort order is determined using the following rules:
+//  - for sam.QueryName the LessByName sam.Record method is used.
+//  - for sam.Coordinate the LessByPostion sam.Record method is used.
+//  - for sam.Unsorted the reader streams are concatenated.
+//  - for sam.Unknown the provided less function is used - if nil
+//    this is the same as sam.Unsorted.
+// For all sort orders other than sam.Unknown, the less parameter
```

Comment:
I would put this above the previous block so that users can skip that section if they know they have unknown.